### ISSUE TITLE
Bug: [no-unnecessary-type-assertion] false positive for string literals

### ISSUE THREAD

**Issue by TkDodo** [typescript-eslint/typescript-eslint#10723]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.7.3&fileType=.tsx&code=MYewdgzgLgBAtgTwHIFc4CMCmAnGBeGAJgFYBYAKFEljgEMAPfGATgAYKKroYAHHYTGFgEABgBIA3gAoAsrSgALAHRwAlmCmJUGHABp4DAJSGYAehgBGVqwC%2BAUhExaEGOIlg0WbPZEBuIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QeoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
const myNumber = 25
const max = 90

const percent = `${(Math.min(myNumber, max)) / 100}%` as `${number}%`;
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

There should be no error in the provided playground

### Actual Result

There’s an error in line 4:

```
This assertion is unnecessary since it does not change the type of the expression. 4:17 - 4:70
```

### Additional Info

Without the type assertion, `percent` would just be inferred as `string`, but we want it to be `${number}%`.

While it’s true that `as const` would give us the correct inference here, as `${number}%` should be allowed, too.

**Comment by ronami** [comment]
That's an interesting issue, I think it's very similar (or even a duplicate) of https://github.com/typescript-eslint/typescript-eslint/issues/10257.

To my understanding, the type [doesn't qualify as a literal](https://github.com/typescript-eslint/typescript-eslint/blob/94e809808a445dde05ab6f284dec010269482eb5/packages/eslint-plugin/src/rules/no-unnecessary-type-assertion.ts#L227) and gets checked as a non-literal, even though the rule has [special logic for template literals](https://github.com/typescript-eslint/typescript-eslint/blob/94e809808a445dde05ab6f284dec010269482eb5/packages/eslint-plugin/src/rules/no-unnecessary-type-assertion.ts#L165-L167). If this used `as const` instead, I think the rule will treat it as a literal (as the asserted type is checked for being a literal) and this should pass ([playground link](https://typescript-eslint.io/play/#ts=5.7.3&fileType=.tsx&code=MYewdgzgLgBAtgTwHIFc4CMCmAnGBeGAJgFYBYAKFEljgEMAPfGATgAYKKroYAHHYTGFgEABgBIA3gAoAsrSgALAHRwAlmCmJUGHABp4DAJSGYAehgBGVqwC%2BAUhExaEGFygBuIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QeoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)).

I've written some thoughts about this (along with tackling this particular case) in https://github.com/typescript-eslint/typescript-eslint/pull/10631.

**Comment by JoshuaKGoldberg** [comment]
Agreed, this is a valid bug and also a subset of #10257. Fun!
